pika/connection.py

```diff
@@ -1296,8 +1327,21 @@ def _on_channel_cleanup(self, channel):
         except KeyError:
             LOGGER.error('Channel %r not in channels',
                          channel.channel_number)
-        if self.is_closing and not self._has_open_channels:
-            self._on_close_ready()
+        if self.is_closing:
```

Comment:
REVIEWME logic in this block makes sure that all channels completed closing before continuing with closing of the connection.